### PR TITLE
power-profiles-daemon: init at 0.1, add service module

### PR DESCRIPTION
**Pull request by mvnetbiz**
###### Motivation for this change
Adds user selected power profile handling through DBus daemon.
nixpkgs issue  #97714

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mvnetbiz** [comment]
@jtojnar It doesn't seem to build man pages, but I think that is broken upstream. It also looks like it builds gtk-doc but that directory gets removed in the fixup phase? Maybe there is something wrong with the expression, I don't know how to debug that.

**Comment by jtojnar** [comment]
Multiple-outputs hook moves gtk-doc to `devdoc` output, or to `/dev/null` if the former is not available. Add `outputs = [ "out" "devdoc" ];` between version and `src` attributes.

It does not look like there are man pages, the command for generating them is not invoked by anything: https://gitlab.freedesktop.org/hadess/power-profiles-daemon/-/blob/7b5321131a7516e1e309f48357fffa2ae909863d/docs/meson.build#L48-60

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  src = fetchFromGitLab {
+    domain = "gitlab.freedesktop.org";
+    owner = "hadess";
+    repo = "power-profiles-daemon";
+    rev = version;
+    sha256 = "012w3aryw5d43dr9jj5i6wy2a0n21jidr4ggs9ix7d4z9byr175w";
+  };
+
+  nativeBuildInputs = [
+    pkgconfig
```
Comment (jtojnar) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:32]: Drop some aliases:

```suggestion
    pkg-config
```

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    meson
+    ninja
+    gettext
+    gtk-doc
+    docbook_xsl
```
Comment (jtojnar) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:37]:
```suggestion
    docbook-xsl-nons
```

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  ];
+
+  mesonFlags = [
+    "-Dsystemdsystemunitdir=./lib/systemd/system"
+    "-Dgtk_doc=true"
+  ];
+
+  postPatch = ''
+    substituteInPlace docs/meson.build \
+      --replace "http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl" "${docbook_xsl}/xml/xsl/docbook/manpages/docbook.xsl"
```
Comment (jtojnar) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:58]: This should not be necessary, it should be picked up automatically from `$XMLCATALOG` variable set by `findXmlPackages` setup hook.

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  buildInputs = [
+    libgudev
+    systemd
+    upower
+    glib
+  ];
+
+  mesonFlags = [
+    "-Dsystemdsystemunitdir=./lib/systemd/system"
```
Comment (jtojnar) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:52]:
```suggestion
    "-Dsystemdsystemunitdir=${placeholder "out"}/lib/systemd/system"
```

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  postPatch = ''
+    substituteInPlace docs/meson.build \
+      --replace "http://docbook.sourceforge.net/release/xsl/current/manpages/docbook.xsl" "${docbook_xsl}/xml/xsl/docbook/manpages/docbook.xsl"
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = "https://gitlab.freedesktop.org/hadess/power-profiles-daemon";
+    description = "Makes user-selected power profiles handling available over D-Bus";
+    platforms = platforms.linux;
+    license = licenses.gpl3;
```
Comment (jtojnar) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:65]: `gpl3` is unclear, use `gpl3Plus` or `gpl3Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

**Comment by mvnetbiz** [comment]
Thanks. I just did these fixes locally, I didn't know about GitHub's "Commit suggestion," it looks like it credits the commenter, is that the preferred approach? Also, this PR should wait until integration into gnome shell is released, since there is no UI for it right?

**Comment by jtojnar** [comment]
For trivial fixes like these, attribution is not necessary (and the extra commits will be squashed onto the original commit anyway).

Merging this is fine just for coverage. And it might be usefult if someone wants to control this using D-Bus or by applying the g-c-c patches to their system.

**Comment by mvnetbiz** [comment]
Added nixos test, should I squash commits?

**Comment by jtojnar** [comment]
I would say either is fine. As long as the individual commits evaluate they are okay.

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    upower
+    glib
+  ];
+
+  mesonFlags = [
+    "-Dsystemdsystemunitdir=${placeholder "out"}/lib/systemd/system"
+    "-Dgtk_doc=true"
+  ];
+
+  meta = with stdenv.lib; {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:58]:
```suggestion
  meta = with lib; {
```

stdenv.lib got recently deprecated.

File: pkgs/os-specific/linux/power-profiles-daemon/default.nix [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+  nativeBuildInputs = [
+    pkg-config
+    meson
+    ninja
+    gettext
+    gtk-doc
+    docbook-xsl-nons
+    docbook_xml_dtd_412
+    libxml2
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/power-profiles-daemon/default.nix:41]: You are sure about this being in nativeBuildInputs. Looks very strange to me.

Comment (jtojnar) [reply pkgs/os-specific/linux/power-profiles-daemon/default.nix:41]: `libxml2` contains `xmllint`, which is often used for validating DocBook documents. Should be probably commented, though.

**Comment by jtojnar** [comment]
Sorry, I forgot about this, should be good to go. Will merge once the CI succeeds.

**Comment by siraben** [comment]
One occurrence of `stdenv.lib` got through.

Closed by #112507

**Comment by LEXUGE** [comment]
https://github.com/NixOS/nixpkgs/blob/d600f006643e074c2ef1d72e462e218b647a096c/nixos/modules/services/hardware/power-profiles-daemon.nix#L37

I feel this is problematic, because regardless whether user enabled `power-profiles-daemon` or not, if they use `tlp`, the assertion fails

**Comment by jtojnar** [comment]
Should not that assertion being in `mkIf cfg.enable` make it only trigger when user enabled p-p-d?

**Comment by mvnetbiz** [comment]
I'm still using TLP on my laptop since this commit, so must be something else going on if you are having a problem.

**Comment by srid** [comment]
I just hit this,

```
❯ sudo nixos-rebuild switch
warning: Git tree '/etc/nixos' is dirty
warning: updating lock file '/etc/nixos/flake.lock':
* Updated 'nixpkgs': 'github:nixos/nixpkgs/1c16013bd6e94da748b41cc123c6b509a23eb440' -> 'github:nixos/nixpkgs/9468b073267c9159d79b89686fabd09367f74ac3'
warning: Git tree '/etc/nixos' is dirty
building the system configuration...
warning: Git tree '/etc/nixos' is dirty
error:
       Failed assertions:
       - You have set services.power-profiles-daemon.enable = true;
       which conflicts with services.tlp.enable = true;
```

The tlp setting seems to be coming from https://github.com/NixOS/nixos-hardware which I use.

**Pull request merged**